Fix/excess ice cold start

mvdebolskiy opened the pull request:
### Description of changes

Changed the way soil temperature is initialized when excess ice is on and the model starts from cold.

### Specific notes

Added 2 parameters: `excess_ice_coldstart_depth` and `excess_ice_coldstart_temp` which control top depth and soil temperature for soil layers in columns where excess ice is present. Other columns get their default soil temperature.

`excessicestream_type` has been taken out of `waterstate_type` and its routines are called directly in `clm_inst%Init`. Checks for `UseExcessIceStreams()` are still in place in `WaterStateType.F90` for double-checking.

Contributors other than yourself, if any:

CTSM Issues Fixed (include github issue #):

 Fixes #2384 
 Fixes #2373 

Are answers expected to change (and if so in what way)?

Yes. Anything that depends on deep soil temperature will be different.

Any User Interface Changes (namelist or namelist defaults changes)?

Two namelist variables are added to `clm_inparm` group: `excess_ice_coldstart_depth` and `excess_ice_coldstart_temp` . Their default values are 0.5m and -2.15 deg C. 

Testing performed, if any: Ran aux_clm on Derecho some tests on izumi

mvdebolskiy commented on bld/CLMBuildNamelist.pm:
```diff
@@ -4564,14 +4564,18 @@ sub setup_logic_exice {
   my $use_exice = $nl->get_value( 'use_excess_ice' );
   my $use_exice_streams = $nl->get_value( 'use_excess_ice_streams' );
   # IF excess ice streams is on
+  if ( (not defined($use_exice_streams) && value_is_true($use_exice)) && ($nl_flags->{'clm_start_type'} == /cold/  || $nl_flags->{'clm_start_type'} == /arb_ic/ ) ) {
```
Not quite sure these are sufficient. Might need to use `finidat == ''` in the condition.

ekluzek replied:
Logic was updated to use finidat, which is likely more robust.

ekluzek replied:
@mvdebolskiy The one thing here is that it looks like potentially the user might set use_excess_ice_streams one way and the logic here might end up overriding what they did. It's better to only set it if the user didn't touch it, and abort with an error if the user tried to do something nonsensical.

ekluzek replied:
OK, I adjusted a few things and think it's good now.

ekluzek replied:
@mvdebolskiy I adjusted the namelist a bit, feel free to look it over and give feedback. And if you'd like to meet we can do that.

ekluzek replied:
Resolving this for now so we'll be able to merge when needed.

mvdebolskiy commented:
Also #2444 might be related

mvdebolskiy commented:
Here are results from I2000Clm60BgcCrop simulations from coldstart:
Given that the forcing gets stuck at 2015 for the rest of the run, it seems reasonable, though the initial temperatures can be lowered a little.

![exice2000](https://github.com/user-attachments/assets/039f8266-961d-4a7f-add9-f176004b728c)
![tsoi2000](https://github.com/user-attachments/assets/c4bfca09-f371-4742-916e-a706a27ba585)
![exice2100](https://github.com/user-attachments/assets/373a0321-f66a-44ed-ab45-0ec6eaae737b)
![tsoi2100](https://github.com/user-attachments/assets/8f53eda6-aa20-452a-9b74-0df7b1fce6c1)

mvdebolskiy commented:
- [ ] add readNamelist subroutine into TemperatureMod and get rid of global clm_varctl declarations.
 - [ ] Make a test where finidat has no excess ice but use_excess_ice is true and use_excess_ice_streams = false.
 - [ ] Add endrun to WaterState, so the model stops if there is excess ice on the restart file but the use_excess_ice is set to false.

ekluzek commented on src/biogeophys/TemperatureType.F90:
```diff
@@ -141,7 +141,7 @@ module TemperatureType
   !------------------------------------------------------------------------
   subroutine Init(this, bounds, &
```
As we want to read the namelist here, you'll need to add NLFilename to the arguments, here and in the call.

ekluzek commented on src/biogeophys/TemperatureType.F90:
```diff
@@ -156,6 +156,7 @@ subroutine Init(this, bounds, &
     logical           , intent(in) :: is_prog_buildtemp    ! Prognostic building temp is being used
+    real(r8)          , intent(in) :: exice_init_stream_col(bounds%begc:)  ! initial excess ice concentration from the stream file
 
     call this%InitAllocate ( bounds )
```
Add a call to a new ReadNML subroutine

``` fortran
    call this%ReadNML( NLFilename )
```

ekluzek commented on src/biogeophys/WaterType.F90:
```diff
@@ -240,14 +241,14 @@ subroutine Init(this, bounds, NLFilename, &
   subroutine InitForTesting(this, bounds, params, &
        h2osno_col, snow_depth_col, watsat_col, &
-       t_soisno_col, use_aquifer_layer, NLFilename)
+       t_soisno_col, use_aquifer_layer, NLFilename, exice_init_stream_col)
```
This change, likely means the PF unit tests will need to change. But, I wonder if it's really needed there...

ekluzek replied:
Just tested and they are working, so this is good.

ekluzek commented on src/main/controlMod.F90:
```diff
@@ -256,7 +256,8 @@ subroutine control_init(dtime)
 
-    namelist /clm_inparm/ use_excess_ice
+    ! excess ice flag and parameters
+    namelist /clm_inparm/ use_excess_ice , excess_ice_coldstart_depth, excess_ice_coldstart_temp
```
As above more to TemperatureType.F90, as well as the changes below.

ekluzek commented on src/biogeophys/TemperatureType.F90:
```diff
@@ -156,6 +156,7 @@ subroutine Init(this, bounds, &
     real(r8)          , intent(in) :: em_perroad_lun(bounds%begl:)
     logical           , intent(in) :: is_simple_buildtemp  ! Simple building temp is being used
     logical           , intent(in) :: is_prog_buildtemp    ! Prognostic building temp is being used
+    real(r8)          , intent(in) :: exice_init_stream_col(bounds%begc:)  ! initial excess ice concentration from the stream file
```
The one thing I wonder about though is the name -- I don't get a nice sense of what quantity it is by the name. Maybe include concentration (or conc) in the name? 

I'd also suggest removing "stream" from the name as it will still be set to something without streams.

ekluzek replied:
Changed to exice_init_conc_col

ekluzek reviewed (changes requested):
Thanks for going over the code with me @mvdebolskiy. Good to see this come in so that we can turn excess-ice on by default for clm6_0 physics. And to have default finidat datasets that have spunup excess-ice in them.

I ask for some changes, some of which we already talked about. And some that probably just appear here:
- [ ] The changes you already talked about...
- [ ] Figure out a name for the new namelist that will be in TemperatureType
- [ ] Add a ReadNML subroutine to TemperatureType, with the new namelist variables in it.
- [ ] Make sure PF unit tests work (do what the README file in the "src" directory says on Derecho
- [x] Check the values of the new namelist items to make sure they are valid
- [ ] Think about changing the name of exice_init_stream_col
- [ ] There's a couple formatting and documentation suggestions I have

These should all be reasonable. Let me know if you'd like to meet to go over how to add the new namelist in. But, there are lots of examples so that should be straightforward to do.

ekluzek commented:
@mvdebolskiy and I talked about the procedure to bring this in. He's not up to making tags himself at this point. So I'll work with him on it. Right now he's going to run some excess ice tests on Derecho to make sure it's working correctly. Then I'll run aux_clm tests for Derecho and Izumi and point him to issues to fix.

This does change answers. But, we actually could make it bit-for-bit by setting the new namelist values appropriately. So that should be something we think about.

I do want to teach him to run aux_clm and create baselines on Derecho. But, we'll take that one step at a time...

mvdebolskiy commented:
> * [ ]  Figure out a name for the new namelist that will be in TemperatureType

    Added  `clm_temperature_inparm` group (for future handling of other coldstart temperature vars)
> * [ ]  Add a ReadNML subroutine to TemperatureType, with the new namelist variables in it.

 Added. Made NLFilename to be optional for unittests to be happy.
> * [ ]  Make sure PF unit tests work (do what the README file in the "src" directory says on Derecho
 Ran `../cime/scripts/fortran_unit_testing/run_tests.py --build-dir unit_tests.temp` : `100% tests passed, 0 tests failed out of 52`

> * [ ]  Check the values of the new namelist items to make sure they are valid

Checked both in `CLMBuildNamelist.pm` and in `TemperatureType%InitCold`
> * [ ]  Think about changing the name of exice_init_stream_col

Changed to `excie_init_conc_col`


Added a warning (instead of endrun) when starting with `use_excess_ice=.false.` from initial conditions that have it.
Note, simple `flag=read` would not work, since there is a call with `write` and `define` before (for init_interp) purposes, though the `RestartExcessIceIssue` helped.

New baseline on izumi: `/home/mvdebolskiy/bl/fix-excess-ice/ERS_D.f10_f10_mg37.I1850Clm60Sp.izumi_nag.clm-ExcessIceStreams`

ekluzek commented on bld/CLMBuildNamelist.pm:
```diff
@@ -4581,15 +4581,25 @@ sub setup_logic_exice {
+   # if excess ice is turned off
+   } elsif ( (not defined($use_exice_streams)) && (not value_is_true($use_exice)) ) {
+   $nl->set_variable_value('exice_streams', 'use_excess_ice_streams' , '.false.');
+   $use_exice_streams = '.false.';
```
Brought in.

ekluzek commented:
@mvdebolskiy I'm planning on doing the following things to finish out the tag. Just FYI, unless you want to comment or contribute:

- [ ] Add some build-namelist unit tests for the new excess ice error conditions
- [ ] Make sure the build-namelist changes are solid, if not and easy go ahead and fix, otherwise file an issue
- [ ] Make sure we have the needed tests on Derecho for excess ice, if not add one
- [ ] Update to ctsm5.2.017
- [ ] Close all the conversations
- [ ] Take a stab at creating the ChangeLog for ctsm5.2.018 (have @mvdebolskiy review and vet)
- [ ] Run baselines for Derecho and Izumi for ctsm5.2.018
- [ ] Finalize the tag as normal

If any problems crop up, I'll consult with you and get a fix in and/or file an issue to fix later.

ekluzek commented on doc/ChangeLog:
```diff
@@ -1,4 +1,108 @@
 ===============================================================
+Tag name: ctsm5.2.018
+Originator(s): mvdebolskiy
+Date: Wed 31 Jul 2024 04:38:36 PM MDT
```
@mvdebolskiy this is a start at the ChangeLog please review and lets make any important changes needed.

ekluzek commented:
Baselines look as expected on Derecho and Izumi. Izumi shows differences for ERS_D.f10_f10_mg37.I1850Clm60Sp.izumi_nag.clm-ExcessIceStreams that are significant, and this is without a cold-start. Which we think shouldn't show differences. So we are investigating. 

I'll also run SMS_Lm12.f09_f09_mg17.I1850Clm60Sp.derecho_intel.clm-ExcessIceStartup_output_sp_exice on Derecho, creating a baseline for ctsm5.2.017 to check what happens there.
